File: lab-4/src/com/company/things/Door.java
package com.company.things;

import abstracts.Closable;
import abstracts.Thing;
import com.company.Quality;
import com.company.Sound;
import types.Color;

public class Door extends Thing implements Closable {

    public Door(Sound sound, Quality quality) {
        super("Дверь", sound, quality, Color.UNDEFINED);
    }

    @Override
    public void close() {
        class Knock extends Sound {
            public Knock(){
                super("Стук");
            }
        }

        Knock knock = new Knock();

        System.out.print("Дверь " + color + "цвета закрылась. ");
        knock.sound();
    }
}

File: lab-4-kotlin/src/com/company/things/Door.kt
package com.company.things

import com.company.Sound
import com.company.Quality
import abstracts.Thing
import abstracts.Closable
import types.Color

class Door(sound: Sound, quality: Quality) :
    Thing("Дверь", sound, quality, Color.UNDEFINED), Closable {
    override fun close() {
        class Knock : Sound("Стук")

        val knock = Knock()
        print("Дверь " + color + "цвета закрылась. ")
        knock.sound()
    }
}